L67-cpp-callback-api.md

```diff
@@ -0,0 +1,125 @@
+   - `OnCancel`: for server RPCs only, called if an RPC is canceled before it has a chance to successfully send status to the client side. Should be used for any cleanup associated with cancellation. Note that even cancelled RPCs _must_ get a `Finish` call to cleanup all their library state. *Server only*
+   - `OnDone`: called when all outstanding RPC operations are completed for both client and server RPCs. For the client-side, it additionally provides the status that the server sent with its `Finish` call. Should be used for any application-level RPC-specific cleanup. *Client and server.*
+   - _Thread safety_: the above calls may take place concurrently, except that `OnDone` will always take place after all other reactions. No further RPC operations are permitted to be issued after `OnDone` is invoked.
+   - **IMPORTANT USAGE NOTE** : code in any reaction must not block for an arbitrary amount of time since reactions are executed on a finite-sized, library-controlled threadpool. If any long-term blocking operations (like sleeps, file I/O, synchronous RPCs, or waiting on a condition variable) must be invoked as part of the application logic, then it is important to push that outside the reaction so that the reaction can complete in a timely fashion. One way of doing this is to push that code to a separate application-controlled thread.
+1. RPC completion prevention methods. These are methods provided by the class and are not virtual.
+   - `AddHold`, `RemoveHold`: for client RPCs, prevents the RPC from being considered complete (ready for `OnDone`) until each `AddHold` is matched to a corresponding `RemoveHold`. These are used to perform _extra-reaction flows_, which refers to  streaming operations initiated from outside a reaction method or method handler. *Client only*. (These are not needed at the server because even if there are extra-reaction flows, the end of the application's involvement with a server RPC is very clearly documented by the invocation of the `Finish` call.)
+
+Examples are provided in [the PR to de-experimentalize the callback API](https://github.com/grpc/grpc/pull/25728).
+
+### Unary RPC shortcuts
+
+As a shortcut, client-side unary RPCs _may_ bypass the reactor model by directly providing a `std::function` for the library to call at completion rather than a reactor object pointer. This is semantically equivalent to a reactor in which the `OnDone` function simply invokes the specified function (but can be implemented in a slightly faster way since such an RPC will definitely not wait separately for initial metadata from the server). In practice, this is the common and recommended model for client-side unary RPCs, unless they have a specific need to wait for initial metadata before getting their full response message. As in the reactor model, the function provided as a callback may not include operations that block for an arbitrary amount of time.
```

Comment:
What API is this callback passed to?

Reply:
To the stub call itself. I guess I should give the example more clearly. It's in the examples code in the de-experimentalization PR linked.